src/Generating/FinishGen.cpp

```diff
@@ -2185,3 +2213,82 @@ void cFinishGenOrePockets::imprintSphere(
 
 
 
+cFinishGenForestRocks::cFinishGenForestRocks(int a_Seed, cIniFile & a_IniFile) : m_Noise(a_Seed)
+{
+}
+
+
+
+
+
+void cFinishGenForestRocks::GenFinish(cChunkDesc & a_ChunkDesc)
+{
+	auto Biome = a_ChunkDesc.GetBiome(cChunkDef::Width / 2, cChunkDef::Width / 2);
+	if ((Biome != biMegaTaiga) && (Biome != biMegaTaigaHills))
+	{
+		return;
+	}
+
+	// Choose random position in chunk and place boulder around it
+	auto Pos = Vector3i(
+		m_Noise.IntNoise2DInt(a_ChunkDesc.GetChunkX(), a_ChunkDesc.GetChunkZ()) % cChunkDef::Width,
+		0,
+		m_Noise.IntNoise2DInt(a_ChunkDesc.GetChunkX(), a_ChunkDesc.GetChunkZ()) % cChunkDef::Width
+		);
+	Pos.y = a_ChunkDesc.GetHeight(Pos.x, Pos.z) % cChunkDef::Height;
+
+	if (!cChunkDef::IsValidRelPos(Pos))
+	{
+		return;
+	}
+
+	// Determines the size of the boulder
+	const int TwoLimit = 70;
+	const int ThreeLimit = 90;
+
+	auto RadiusChance = m_Noise.IntNoise2DInt(a_ChunkDesc.GetChunkX(), a_ChunkDesc.GetChunkZ()) % 100;
+	int Radius = 1;
+	if (RadiusChance > TwoLimit && RadiusChance <= ThreeLimit)
+	{
+		Radius = 2;
+	}
+	else if (RadiusChance > ThreeLimit)
+	{
+		Radius = 3;
+	}
+
+	Pos.x = Clamp(Pos.x, 0 + Radius, cChunkDef::Width - Radius - 1);
+	Pos.z = Clamp(Pos.z, 0 + Radius, cChunkDef::Width - Radius - 1);
+
+	auto StartBlock = a_ChunkDesc.GetBlockType(Pos.x, Pos.y + Radius, Pos.z);
+	while (!((StartBlock == E_BLOCK_DIRT) || (StartBlock == E_BLOCK_GRASS)))
+	{
+		Pos.y -= 1;
+		if (!cChunkDef::IsValidRelPos(Pos.addedY(-Radius)))
+		{
+			return;
+		}
+		StartBlock = a_ChunkDesc.GetBlockType(Pos.x, Pos.y + Radius, Pos.z);
+	}
```

Comment:
GetBlockType at the start might be called with out of bounds Y if Pos.y is high enough.

If I understand correctly, this always sinks the tip of the boulder flush with the ground. What do you think about placing the boulder's centre at the first valid StartBlock, then randomly moving down by [0, -Radius], and finally clamping Pos.y?,

Reply:
The thing is that the position might be on top of a tree (the same reason why we don't get grass generation beneath trees) so we would need to move this down to the ground first and then randomize